addons/crm/models/crm_lead.py

```diff
@@ -1011,27 +1012,109 @@ def _get_rainbowman_message(self):
+                meeting_dts.append((meeting.get('start'), meeting.get('stop')))
+            else:
+                meeting_dts.append((meeting.get('start').astimezone(user_pytz).replace(tzinfo=None),
+                                   meeting.get('stop').astimezone(user_pytz).replace(tzinfo=None)))
+
+        # If there are meetings that are still ongoing or to come, only take those.
+        unfinished_meeting_dts = [meeting_dt for meeting_dt in meeting_dts if meeting_dt[1] >= now_dt]
+        relevant_meeting_dts = unfinished_meeting_dts if unfinished_meeting_dts else meeting_dts
+        relevant_meeting_count = len(relevant_meeting_dts)
+
+        if relevant_meeting_count == 1:
+            return "week", relevant_meeting_dts[0][0].date()
+        else:
+            # Range of meetings
+            earliest_start_dt = min(relevant_meeting_dt[0] for relevant_meeting_dt in relevant_meeting_dts)
+            latest_stop_dt = max(relevant_meeting_dt[1] for relevant_meeting_dt in relevant_meeting_dts)
+
+            # The week start day depends on language. We fetch the week_start of user's language. 1 is monday.
+            lang_week_start = self.env["res.lang"].search_read([('code', '=', self.env.user.lang)], ['week_start'])
+            # We substract one to make week_start_index range 0-6 instead of 1-7
+            week_start_index = int(lang_week_start[0].get('week_start', '1')) - 1
+
+            # We compute the weekday of earliest_start_dt according to week_start_index. earliest_start_dt_index will be 0 if we are on the
+            # first day of the week and 6 on the last. weekday() returns 0 for monday and 6 for sunday. For instance, Tuesday in UK is the
+            # third day of the week, so earliest_start_dt_index is 2, and remaining_days_in_week includes tuesday, so it will be 5.
+            # The first term 7 is there to avoid negative left side on the modulo, improving readability.
+            earliest_start_dt_weekday = (7 + earliest_start_dt.weekday() - week_start_index) % 7
+            remaining_days_in_week = 7 - earliest_start_dt_weekday
+
+            # We compute the start of the week following the one containing the start of the first meeting.
+            next_week_start_date = earliest_start_dt.date() + timedelta(days=remaining_days_in_week)
+
+            # Latest_stop_dt must be before the start of following week. Limit is therefore set at midnight of first day, included.
+            meetings_in_same_week = latest_stop_dt <= datetime(next_week_start_date.year, next_week_start_date.month, next_week_start_date.day, 0, 0, 0)
+
+            if meetings_in_same_week:
+                return "week", earliest_start_dt.date()
+            else:
+                return "month", earliest_start_dt.date()
+
     def action_reschedule_meeting(self):
         self.ensure_one()
-        action = self.action_schedule_meeting()
+        action = self.action_schedule_meeting(smart_calendar=False)
```

Comment:
@dbeguin. Hello! Just a small change I had to make to rebase on damien's code. I just added a parameter, default value true, smart_calendar, that indicates if the view should search and use smart_calendar values. When rescheduling, we should avoid that to avoid time loss and inappropriate mode. Does it sound ok for you? cheers!

Reply:
@msyukyu Hello, see comment above =) I figured reschedule action would always prefer week mode, so I decided not to do useless "optimal parameters view" search effort. Looks ok to you? So basically this does not change anything, of what I have in mind. Cheers!

Reply:
Good idea.